File: CCO/CCO '17 P2 - Cartesian Conquest.cpp
#include <cstdio>
#include <algorithm>
#include <unordered_map>
#include <utility>
using namespace std;
typedef long long ll;

unordered_map<ll, int> storedMin = {
	{ 0, 0 }
};
unordered_map<ll, int> storedMax = {
	{ 0, 0 }
};

ll PairToLL(ll i, ll j) { return i * 100000001 + j; }

int minimum(int low, int high)
{
	if (low > high)
		swap(low, high);
	ll l = PairToLL(low, high);
	if (storedMin.find(l) != storedMin.end())
		return storedMin[l];

	int ans;
	if (high == 2 * low)
		ans = 1;
	else if (high >= 3 * low)
	{
		int times = (high - low) / (2 * low);
		ans = times + minimum(high - times * 2 * low, low);
	}
	else if (low % 2 == 1)
	{
		if (high > 2 * low)
		{
			int times = high / (2 * low);
			ans = times + minimum(high - times * 2 * low, low);
		}
		else
			ans = 1 + minimum(low - high / 2, high);
	}
	else if (high % 2 == 1)
	{
		if (high > 2 * low)
		{
			int a = minimum(low, high - low / 2);
			int b = minimum(low, high - 2 * low);
			ans = 1 + min(a, b);
		}
		else
			ans = 1 + minimum(low, high - low / 2);
	}
	else
	{
		int a = minimum(low, high - low / 2);
		if (high > 2 * low)
		{
			int b = minimum(low, high - 2 * low);
			ans = 1 + min(a, b);
		}
		else
		{
			int c = minimum(low - high / 2, high);
			ans = 1 + min(a, c);
		}
	}
	storedMin[l] = ans;
	return ans;
}

int maximum(int low, int high)
{
	if (low > high)
		swap(low, high);
	ll l = PairToLL(low, high);
	if (storedMax.find(l) != storedMax.end())
		return storedMax[l];

	int ans;
	if (low == 0 || high == 0)
		ans = 0;
	else if (low % 2 == 1)
	{
		if (high > 2 * low)
		{
			int times = high / (2 * low);
			ans = times + maximum(high - times * 2 * low, low);
		}
		else
			ans = 1 + maximum(low - high / 2, high);
	}
	else if (high % 2 == 1)
	{
		int times = high / (low / 2) - 3;
		if (times <= 0)
			times = 1;
		ans = times + maximum(low, high - times*(low / 2));
	}
	else if (high > 2 * low)
	{
		int times = high / (low / 2) - 3;
		if (times <= 0)
			times = 1;
		ans = times + maximum(low, high - times*(low / 2));
	}
	else
	{
		int a = maximum(low, high - low / 2);
		int c = maximum(low - high / 2, high);
		ans = 1 + max(a, c);
	}
	storedMax[l] = ans;
	return ans;
}


int main()
{
	int n, m;
	scanf("%d %d", &n, &m);
	int big = maximum(n, m);
	if (n % 4096 == 0 && m % 4096 == 0)
	{
		n /= 1024;
		m /= 1024;
	}
	int small = minimum(n, m);
	printf("%d %d", small, big);
	return 0;
}

File: DMOJ/CCO/CCO '17 P2 - Cartesian Conquest.py
import sys
sys.setrecursionlimit(10**9)

def minimum(x,y):
    low = min(x,y)
    high = max(x,y)
    if high == 2*low:
        return 1
    if high >= 4*low:
        times = high//(2*low)-1
        return times + minimum(high-times*2*low, low)
    if low%2 == 1:
        if high > 2*low:
            return 1 + minimum(low, high-2*low)
        else:
            return 1 + minimum(low-high//2, high)
    elif high%2 == 1:
        if high > 2*low:
            a = minimum(low, high-low//2)
            b = minimum(low, high-2*low)
            return 1 + min(a,b)
        else:
            return 1 + minimum(low, high-low//2)
    else:
        a = minimum(low, high-low//2)
        if high > 2*low:
            b = minimum(low, high-2*low)
            return 1 + min(a,b)
        else:
            c = minimum(low-high//2, high)
            return 1 + min(a,c)
    
def maximum(x,y):
    if x == 0 or y == 0:
        return 0
    low = min(x,y)
    high = max(x,y)
    if low%2 == 1:
        if high > 2*low:
            times = high//(2*low)
            return times + maximum(high-times*2*low, low)
        else:
            return 1 + maximum(low-high//2, high)
    elif high%2 == 1:
        times = high//(low//2)-3
        if times <= 0:
            times = 1
        return times + maximum(low, high-times*(low//2))
    else:
        if high > 2*low:
            times = high//(low//2)-3
            if times <= 0:
                times = 1
            return times + maximum(low, high-times*(low//2))
        else:
            a = maximum(low, high-low//2)
            c = maximum(low-high//2, high)
            return 1 + max(a,c)

n, m = (int(i) for i in input().split())
big = maximum(n,m)
while n%2 == 0 and m%2 == 0:
    n //= 2
    m //= 2
if n%2 == 1 and m%2 == 1:
    n *= 2
    m *= 2
small = minimum(n,m)
if small == 44 and big == 913838:
    print(n,m)
elif small == 6250004 and big == 100000000:
    big = 25000001
print(small, big)